libs/pika/async_mpi/include/pika/async_mpi/transform_mpi.hpp

```diff
@@ -336,100 +439,306 @@ namespace pika::mpi::experimental {
             {
-                return operation_state<Receiver>(
-                    PIKA_FORWARD(Receiver, receiver), s.f, s.sender, s.stream);
+                return operation_state<Receiver>(PIKA_FORWARD(Receiver, receiver), s.f, s.sender, s.stream);
             }
 
             template <typename Receiver>
-            friend constexpr auto
-            tag_invoke(pika::execution::experimental::connect_t,
-                transform_mpi_sender_type&& s, Receiver&& receiver)
+            friend constexpr auto tag_invoke(
+                pika::execution::experimental::connect_t, transform_mpi_sender_type&& s, Receiver&& receiver)
             {
                 return operation_state<Receiver>(
-                    PIKA_FORWARD(Receiver, receiver), PIKA_MOVE(s.f),
-                    PIKA_MOVE(s.sender), s.stream);
+                    PIKA_FORWARD(Receiver, receiver), PIKA_MOVE(s.f), PIKA_MOVE(s.sender), s.stream);
             }
         };
     }    // namespace transform_mpi_detail
 
-    inline constexpr struct transform_mpi_t final
-      : pika::functional::detail::tag_fallback<transform_mpi_t>
+    inline constexpr struct transform_mpi_t final : pika::functional::detail::tag_fallback<transform_mpi_t>
     {
     private:
         template <typename Sender, typename F,
-            PIKA_CONCEPT_REQUIRES_(
-                pika::execution::experimental::is_sender_v<Sender>)>
+            PIKA_CONCEPT_REQUIRES_(pika::execution::experimental::is_sender_v<Sender>)>
         friend constexpr PIKA_FORCEINLINE auto
-        tag_fallback_invoke(transform_mpi_t, Sender&& sender, F&& f,
-            mpi::experimental::stream_type s =
-                mpi::experimental::stream_type::automatic)
+        tag_fallback_invoke(transform_mpi_t, Sender&& sender, F&& f, stream_type s = stream_type::automatic)
         {
-            return transform_mpi_detail::transform_mpi_sender<Sender, F>{
-                PIKA_FORWARD(Sender, sender), PIKA_FORWARD(F, f), s};
+            using namespace transform_mpi_detail;
+            PIKA_DP(mpi_tran,
+                debug(debug::detail::str<>("tag_fallback_invoke"), "stream", detail::stream_name(s)));
+
+            if constexpr (pika::execution::experimental::detail::has_completion_scheduler_v<
+                              pika::execution::experimental::set_value_t, std::decay_t<Sender>>)
+            {
+                return pika::execution::experimental::transfer(
+                    transform_mpi_sender<Sender, F>{PIKA_FORWARD(Sender, sender), PIKA_FORWARD(F, f), s},
+                    pika::execution::experimental::get_completion_scheduler<
+                        pika::execution::experimental::set_value_t>(sender));
+            }
+            else
+            {
+                namespace ex = pika::execution::experimental;
+                namespace mpi = pika::mpi::experimental;
+                auto mode = mpi::get_completion_mode();
+                if (mode == 0)
```

Comment:
Could we name the modes (with an enum) instead?

Reply:
yes, an enum could be used, but then the ENV var setting would probably need to be changed to use string or something. Not sure it makes anything better yet. Same comment about `experimental` features applies here.

Reply:
Fair enough, let's keep it like this for now.